tests/hardhat/Comptroller/Diamond/comptrollerTest.ts

```diff
@@ -534,5 +592,144 @@ describe("Comptroller", () => {
+          const errCode = await comptroller.callStatic.liquidateBorrowAllowed(
+            vToken.address,
+            vToken.address,
+            accounts[0].address,
+            root.address,
+            convertToUnit("1", 18),
+          );
+          expect(errCode).to.equal(0);
+        });
+
+        it("allows to repay 100% of the borrow", async () => {
+          vToken.borrowBalanceStored.returns(convertToUnit("1", 18));
+          const errCode = await comptroller.callStatic.liquidateBorrowAllowed(
+            vToken.address,
+            vToken.address,
+            accounts[0].address,
+            root.address,
+            convertToUnit("1", 18),
+          );
+          expect(errCode).to.equal(0);
+        });
+
+        it("fails with TOO_MUCH_REPAY if trying to repay > borrowed amount", async () => {
+          vToken.borrowBalanceStored.returns(convertToUnit("0.99", 18));
+          const errCode = await comptroller.callStatic.liquidateBorrowAllowed(
+            vToken.address,
+            vToken.address,
+            accounts[0].address,
+            root.address,
+            convertToUnit("1", 18),
+          );
+          expect(errCode).to.equal(17);
+        });
+
+        it("checks the shortfall if isForcedLiquidationEnabled is set back to false", async () => {
+          await comptroller._setForcedLiquidation(vToken.address, false);
+          vToken.borrowBalanceStored.returns(convertToUnit("100", 18));
+          const errCode = await comptroller.callStatic.liquidateBorrowAllowed(
+            vToken.address,
+            vToken.address,
+            accounts[0].address,
+            root.address,
+            convertToUnit("1", 18),
+          );
+          expect(errCode).to.equal(3);
+        });
+      });
+
+      describe("isForcedLiquidationEnabled == false", async () => {
+        let comptrollerLens: FakeContract<ComptrollerLens>;
+
+        beforeEach(async () => {
+          comptrollerLens = await smock.fake<ComptrollerLens>("ComptrollerLens");
+          await comptroller._setComptrollerLens(comptrollerLens.address);
+          await comptroller._setCloseFactor(convertToUnit("0.5", 18));
+        });
+
+        generalTests();
+
+        it("fails if borrower has 0 shortfall", async () => {
+          vToken.borrowBalanceStored.returns(convertToUnit("100", 18));
+          comptrollerLens.getHypotheticalAccountLiquidity.returns([0, 1, 0]);
+          const errCode = await comptroller.callStatic.liquidateBorrowAllowed(
+            vToken.address,
+            vToken.address,
+            accounts[0].address,
+            root.address,
+            convertToUnit("1", 18),
+          );
+          expect(errCode).to.equal(3);
+        });
+
+        it("succeeds if borrower has nonzero shortfall", async () => {
+          vToken.borrowBalanceStored.returns(convertToUnit("100", 18));
+          comptrollerLens.getHypotheticalAccountLiquidity.returns([0, 0, 1]);
```

Comment:
Shouldn't we remove lens here and fake oracle price instead? AFAIU we don't use lens in Diamond Comptroller anymore

Reply:
we are using lens in the FacetBase, https://github.com/VenusProtocol/venus-protocol/blob/feat/diamond-proxy/contracts/Comptroller/Diamond/facets/FacetBase.sol#L141